paimon-spark/paimon-spark-common/src/main/java/org/apache/paimon/spark/catalog/functions/BucketFunction.java

```diff
@@ -0,0 +1,724 @@
+import org.apache.spark.sql.connector.catalog.functions.UnboundFunction;
+import org.apache.spark.sql.types.BinaryType;
+import org.apache.spark.sql.types.BooleanType;
+import org.apache.spark.sql.types.ByteType;
+import org.apache.spark.sql.types.DataType;
+import org.apache.spark.sql.types.DataTypes;
+import org.apache.spark.sql.types.Decimal;
+import org.apache.spark.sql.types.DecimalType;
+import org.apache.spark.sql.types.DoubleType;
+import org.apache.spark.sql.types.FloatType;
+import org.apache.spark.sql.types.IntegerType;
+import org.apache.spark.sql.types.LongType;
+import org.apache.spark.sql.types.ShortType;
+import org.apache.spark.sql.types.StringType;
+import org.apache.spark.sql.types.StructField;
+import org.apache.spark.sql.types.StructType;
+import org.apache.spark.sql.types.TimestampNTZType;
+import org.apache.spark.sql.types.TimestampType;
+import org.apache.spark.unsafe.types.UTF8String;
+
+import java.io.IOException;
+import java.io.ObjectInputStream;
+import java.lang.reflect.InvocationTargetException;
+import java.util.Arrays;
+import java.util.Map;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+/** A Spark function implementation for the Paimon bucket transform. */
+public class BucketFunction implements UnboundFunction {
+    private static final int NUM_BUCKETS_ORDINAL = 0;
+    private static final int SPARK_TIMESTAMP_PRECISION = 6;
+
+    private static final Map<String, Class<? extends BucketGeneric>> BUCKET_FUNCTIONS;
+
+    static {
+        ImmutableMap.Builder<String, Class<? extends BucketGeneric>> builder =
+                ImmutableMap.builder();
+        builder.put("BucketBoolean", BucketBoolean.class);
+        builder.put("BucketByte", BucketByte.class);
+        builder.put("BucketShort", BucketShort.class);
+        builder.put("BucketInteger", BucketInteger.class);
+        builder.put("BucketLong", BucketLong.class);
+        builder.put("BucketFloat", BucketFloat.class);
+        builder.put("BucketDouble", BucketDouble.class);
+        builder.put("BucketString", BucketString.class);
+        builder.put("BucketDecimal", BucketDecimal.class);
+        builder.put("BucketTimestamp", BucketTimestamp.class);
+        builder.put("BucketBinary", BucketBinary.class);
+
+        // Joint bucket fields of common types
+        builder.put("BucketIntegerInteger", BucketIntegerInteger.class);
+        builder.put("BucketIntegerLong", BucketIntegerLong.class);
+        builder.put("BucketIntegerString", BucketIntegerString.class);
+        builder.put("BucketLongInteger", BucketLongInteger.class);
+        builder.put("BucketLongLong", BucketLongLong.class);
+        builder.put("BucketLongString", BucketLongString.class);
+        builder.put("BucketStringInteger", BucketStringInteger.class);
+        builder.put("BucketStringLong", BucketStringLong.class);
+        builder.put("BucketStringString", BucketStringString.class);
```

Comment:
For primitive types and common composite types, I added separate bucket function classs to improve [performance](https://github.com/apache/spark/blob/1a42aa5bd44e7524bb55463bbd85bea782715834/sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/functions/ScalarFunction.java#L35), while `BucketGeneric` handles other types.

Reply:
Have you tested the performance of this implementation? I think the current approach is too specialized, and the data conversion involving Spark and Paimon doesn't reuse the code.

Reply:
https://github.com/zhongyujiang/incubator-paimon/tree/gh/spark-v2-write_function-benchmark
```text
Benchmark                                            Mode  Cnt  Score   Error  Units
PaimonBucketFunctionBenchmark.bucketFunction           ss    5  0.416 ± 0.033   s/op
PaimonBucketFunctionBenchmark.genericBucketFunction    ss    5  0.685 ± 0.302   s/op
```
Here is a JMH benchmark result. 
I think this optimization has limited impact on write performance, since a significant amount of time is spent on network and compression during data writing. However, there is indeed a noticeable performance difference between the two implementations.
This part of the code can't be reused since the invoke method takes different input types.

Reply:
It looks like the difference is small. I added a benchmark https://github.com/apache/paimon/pull/5418. 
Will test the performance of the existing bucket function and this PR's, including the function and e2e write cost.

Reply:
If we only compare these two function implementations, you can see from the results above that there's a clear difference 0.685 / 0.416 = 1.64.
In real write scenarios, I agree that the impact of this optimization is very minimal.

Reply:
FYI, I’ve already added a benchmark comparison between the v2 write and the previous v1 write [here](https://github.com/apache/paimon/pull/5241/files#diff-b60ea9f4f3114528a9e9c1c4880f61c6c1d2b3bc7c467b0b22b44abe6ad69b71), the results is attached in the PR too. But it didn’t include the newly added v1 bucket function optimization, I think the benchmark results from JMH would be more reliable.

Reply:
Sure, let me update this.